### DISCUSSION TITLE
New collection review: purestorage.fusion

### DISCUSSION THREAD

**Discussion by sdodsley** [ansible-collections/ansible-inclusion#48] [Resolved reviews]
This collection provides support for the Pure Storage Fusion control plane.

Ansible Galaxy: https://galaxy.ansible.com/purestorage/fusion
GitHub: https://github.com/Pure-Storage-Ansible/Fusion-Collection
Issues Tracker: https://github.com/Pure-Storage-Ansible/Fusion-Collection/issues
Is the collection part of Automation Hub: Yes
We meet [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst): Yes

**Comment by Andersson007** [comment]
@sdodsley hello, thank you for the submission! I hope the @ansible-collections/steering-committee will take a look ASAP
Here's the [Collection inclusion checklist](https://github.com/ansible-collections/overview/blob/main/collection_checklist.md).

**Comment by mariolenz** [comment]
# Ansible Collections Checklist (short version)

_For details about the following points, refer to the [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst)._

Every comment should say whether the reviewer expects it to be addressed, or whether it's optional.

**Public availability and communication:**
- [x] published on [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [x] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
- [x] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues
- [x] have a public SCM repository, releases of the collection are tagged in this repository

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
- [x] follows [licensing rules](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#licensing)
- [x] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
- [x] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html); as well as these other requirements:
  - [x] modules that only gather information are named `<something>_info`
  - [x] modules that return `ansible_facts` are named `<something>_facts` and do not return non-facts
  - [x] other modules must not allow querying information using specific `state` option values, or similar mechanisms (like `state=get` or `state=query`).  These features should be moved to `<something>_info` or `<something>_fact` modules.
  - [x] `check_mode` is supported in all `_info` and `_facts` modules
- [x] supports Python 2.6 or greater and Python 3.5 or greater. If it does not, read the [full guidelines](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#python-compatibility) to see if you qualify for an exception and document the unsupported [Python versions](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#ansible-and-python-3) in the collection ``README.md`` and in every module and plugin (or in doc fragments)
    
    _comment:_ I'm not sure if this is a good requirement any more. The collection runs on the controller, not on the managed node. If we add it to Ansible 5+ (ansible-core 2.12+), the minimum Python version for the control node is 3.8.

- [x] only uses the [allowed plugin types](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#modules-plugins) in the `plugins/` directory
- [x] has `README.md`
- [x] documentation, examples, and return sections use FQCNs for the `M(..)` [format macros](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation) when referring to modules
- [x] modules (or plugins) from ansible-core use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module or plugin was added
- [x] FQCNs are used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-core
- [x] public plugins, roles and playbooks do not use files outside of `meta/`, `plugins/`, `roles/`, and `playbooks/`

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0, and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
- [x] has changelog, preferably with `changelogs/changelog.yaml`
- [x] collection repository should not contain any large objects (binaries) comparatively to the current Galaxy tarball size limit of 20 MB. For example, package installers for testing purposes shouldn't be added.
- [x] collection repository should not contain any unnecessary files like, for example, temporary files created by development tools. Temporary files MUST be added to `.gitignore`.

**Tests:**

- [x] passed `ansible-test sanity`
- [x] if `test/sanity/ignore*.txt` exists, it does not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [x] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
- [x] all CI tests run against every pull request
- [x] all CI tests run against a commit that releases the collection; if they don't pass, the collection won't be released
- [x] all CI tests run regularly (nightly, or at least once per week)

**Reply by felixfontein** [reply]
> * [ ]  supports Python 2.6 or greater and Python 3.5 or greater. If it does not, read the [full guidelines](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#python-compatibility) to see if you qualify for an exception and document the unsupported [Python versions](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#ansible-and-python-3) in the collection `README.md` and in every module and plugin (or in doc fragments)
>       _comment:_ I'm not sure if this is a good requirement any more. The collection runs on the controller, not on the managed node. If we add it to Ansible 5+ (ansible-core 2.12+), the minimum Python version for the control node is 3.8.

I don't see why the collection **must** be run on the controller. It can also be run on another host (there might be some company policy that all access to the storage has to be made from a certain machine, for example). So while the common case is to execute these modules on the controller, they can also be executed somewhere else.

The individual modules already document that they require Pyhton >= 3.5 (with the docs fragment), so I guess the README of the collection should also say that.

**Reply by sdodsley** [reply]
Added the Python requirement to the README

**Comment by mariolenz** [comment]
@Andersson007 I've copied the checklist and started to work on it. But I won't promise anything, haven't done this before.

**Comment by mariolenz** [comment]
@sdodsley It looks like version 1.0.1 is available on galaxy, but not tagged in the repository. I think this is a violation of

> have a public SCM repository, releases of the collection are tagged in this repository

**Reply by sdodsley** [reply]
Fixed

**Comment by mariolenz** [comment]
@sdodsley 

> collection repository should not contain any unnecessary files like, for example, temporary files created by development tools. Temporary files MUST be added to `.gitignore`.

Pure-Storage-Ansible/Fusion-Collection#25

**Reply by sdodsley** [reply]
Resolved

**Comment by mariolenz** [comment]
@sdodsley I've found two issues with the examples in the module documentation:

1. Please use the FQCN in fusion_info.py (`purestorage.fusion.fusion_info` instead of `fusion_info`)
2. There's a typo in the second example of fusion_ss.py, `purestorage.fusion.fusion_s` instead of `purestorage.fusion.fusion_ss` (missing 's')

**Reply by sdodsley** [reply]
@mariolenz Both fixed

**Comment by mariolenz** [comment]
@Andersson007 This is my first new collection review, but I should say `purestorage.fusion` satisfies all the requirements. I hope I didn't blunder.

**Comment by felixfontein** [comment]
One thing I noticed when glancing at the collection is that it has `ANSIBLE_METADATA` in the modules. That entry is only needed for Ansible 2.9 (or before) **when having non-default values**, i.e. for "Supported" modules. We recommend to remove that, even though right now the presence of that field doesn't hurt.

Another quick note: in https://github.com/Pure-Storage-Ansible/Fusion-Collection/blob/master/plugins/modules/fusion_api_client.py#L41 `C(...)` should be used instead of `I(...)`, and other places such as https://github.com/Pure-Storage-Ansible/Fusion-Collection/blob/master/plugins/modules/fusion_ss.py#L42 use backticks instead of the allowed formatting identifiers (see https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation).

Also the README says `Ansible >= 2.11`, which should say `ansible-core >= 2.11`.

And a final note for today: the copyright header in modules says `# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)`, while the actual file is called [COPYING.GPLv3](https://github.com/Pure-Storage-Ansible/Fusion-Collection/blob/master/COPYING.GPLv3). So the copyright header should say `# GNU General Public License v3.0+ (see COPYING.GPLv3 or https://www.gnu.org/licenses/gpl-3.0.txt)`.

**Reply by sdodsley** [reply]
@felixfontein I beleive I have addressed all your concerns.

**Comment by Andersson007** [comment]
# Ansible Collections Checklist (short version)

_For details about the following points, refer to the [Collection Requirements](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst)._

Every comment should say whether the reviewer expects it to be addressed, or whether it's optional.

Note for reviewers: If you don't know how to check any of the points below, please ask maintainers of the collection you're reviewing or a [Steering Committee member](https://docs.ansible.com/ansible/devel/community/steering/community_steering_committee.html#current-steering-committee-members) for clarifications in comments of corresponding inclusion discussion.

**Public availability and communication:**
- [x] published on [Ansible Galaxy](https://galaxy.ansible.com) with version 1.0.0 or later
- [x] has a policy of releasing, versioning and deprecation announced to contributors and users in some way
- [x] have a Code of Conduct (CoC) compatible with the [Ansible Code of Conduct (CoC)](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)
- [x] have a publicly available issue tracker that does not require a paid level of service to create an account or view issues
- [x] have a public SCM repository, releases of the collection are tagged in this repository

**Standards and documentation:**
- [x] adheres to [semantic versioning](https://semver.org/)
- [x] follows [licensing rules](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#licensing)
  **COMMENT**: I see 2 license files [COPYING.GPLv3](https://github.com/Pure-Storage-Ansible/Fusion-Collection/blob/master/COPYING.GPLv3) and LICENSE, should one of them be removed?
- [x] follows the [Ansible documentation standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html) and the [style guide](https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#style-guide)
  **MUST FIX:** There's no info if modules support check mode or not. Usually there should be something like `- Supports C(check_mode).` in the `Notes:` section.
  **MUST FIX:** In option descriptions in the DOCUMENTATION block some lines end with periods and some not. Please put periods where missed (See the [doc standards](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html)).
  **MUST FIX:** `If not provided, defaults to C(name)` - we use `I()` for option names. Please recheck if you use proper [markdown](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation). Same for cases like `To clear the username/password pair use I(clear) as the password` as we use `C()` for values.
  **MUST FIX** `- name: collect default set of information` - Examples descriptions must start with a capital letter.
- [x] follows [development conventions](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html); as well as these other requirements:
  - [x] modules that only gather information are named `<something>_info`
  - [x] modules that return `ansible_facts` are named `<something>_facts` and do not return non-facts
  - [x] other modules must not allow querying information using specific `state` option values, or similar mechanisms (like `state=get` or `state=query`).  These features should be moved to `<something>_info` or `<something>_fact` modules.
  - [x] `check_mode` is supported in all `_info` and `_facts` modules
- [x] supports Python 2.6 or greater and Python 3.5 or greater. If it does not, read the [full guidelines](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#python-compatibility) to see if you qualify for an exception and document the unsupported [Python versions](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#ansible-and-python-3) in the collection ``README.md`` and in every module and plugin (or in doc fragments)
- [x] only uses the [allowed plugin types](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#modules-plugins) in the `plugins/` directory
- [x] has `README.md`
- [x] documentation, examples, and return sections use FQCNs for the `M(..)` [format macros](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation) when referring to modules
- [x] modules (or plugins) from ansible-core use `ansible.builtin.` as a FQCN prefix
- [x] documentation and return sections use `version_added:` containing the *collection* version for which an option, module or plugin was added
- [x] FQCNs are used for `extends_documentation_fragment:`, unless you are referring to doc_fragments from ansible-core
- [x] public plugins, roles and playbooks do not use files outside of `meta/`, `plugins/`, `roles/`, and `playbooks/`

**Collection management:**
- [x] `galaxy.yml` having `tags` field set
- [x] collection dependencies must have a lower bound on the version which is at least 1.0.0, and are all part of the `ansible` package
- [x] `meta/runtime.yml` defines the minimal version of Ansible which the collection works with
- [x] has changelog, preferably with `changelogs/changelog.yaml`
  **MUST FIX:** There are no entries for versions v1.0.1 and v1.0.2 in the changelog. Do you use [antsibull-changelog](https://docs.ansible.com/ansible/devel/community/collection_contributors/collection_release_without_branches.html#generating-the-changelog)?
- [x] collection repository should not contain any large objects (binaries) comparatively to the current Galaxy tarball size limit of 20 MB. For example, package installers for testing purposes shouldn't be added.
- [x] collection repository should not contain any unnecessary files like, for example, temporary files created by development tools. Temporary files MUST be added to `.gitignore`.
  **MUST FIX:** There's the `runtime.yml` file in the repo's rood directory, which, i think, can be removed.

**Tests:**

- [x] passed `ansible-test sanity`
- [x] if `test/sanity/ignore*.txt` exists, it MUST not contain error codes listed [here](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#ci-testing)
- [x] has CI tests up and running against each of the "major versions" of `ansible-base`/`ansible-core` that the collection supports
- [x] all CI tests MUST run against every pull request
  **MUST FIX**: In GH Actions tab i see that sanity tests run regularly but in jobs related to PRs i see only `Lint` tests ran. Same for release commits. Also see ssbarnea's [comment](https://github.com/ansible-collections/ansible-inclusion/discussions/48#discussioncomment-3029096).
- [x] all CI tests MUST run regularly (nightly, or at least once per week)
- [x] sanity tests MUST run against a commit that releases the collection; if they don't pass, the collection won't be released
  **MUST FIX**: See my comment above

**Reply by Andersson007** [reply]
@sdodsley while i review the collection, could you please (and other maintainers) subscribe to [news-for-maintainers](https://github.com/ansible-collections/news-for-maintainers) if you're still not?

**Reply by ssbarnea** [reply]
I did a very brief look at the repo and I was surprised to see that on its "lint" job it was only running black. While seeing use of black is good, I would have expected ansible-lint, especially as that is part of upload to galaxy.

**Reply by Andersson007** [reply]
@sdodsley i finished my review. Please take a look at it and @ssbarnea 's [comment](https://github.com/ansible-collections/ansible-inclusion/discussions/48#discussioncomment-3029096).

**Reply by sdodsley** [reply]
@Andersson007 The CI sanity tests run against every PR, including releases. We run two separate jobs - one for ansible sanity and one for Lint for every PR. So I'm not sure what you are referring to in your **MUST FIX** for Tests.
All documentation standards have now been adhered to as required.
Extra `LICENSE` file removed as well as unnecessary `runtime.yaml`
Release 1.0.1 and 1.0.2 did use `antsibull-changlelog` but they were just review changes, so nothing in the actual code. 
@ssbarnea Ansible-lint added to CI as requested.

**Comment by Andersson007** [comment]
@sdodsley thanks for clarifying and fixing the stuff, +1 from me to its inclusion.
I've created a corresponding community topic https://github.com/ansible-community/community-topics/issues/117.
If there are no objections from the Steering Committee, I'll add it to the build file on July 7, 2022 and it'll appear in the package in the next Ansible 6.X release.

**Comment by Andersson007** [comment]
Included https://github.com/ansible-community/ansible-build-data/pull/141, will appear in the next Ansible 6.X release. Thanks everyone!
